File: runtime/ios/Sources/IstmoRuntime/PluginHandler.swift
import Foundation

public protocol PluginHandler: AnyObject {

    func handleCall(instanceId: UInt64, method: String, payload: Data) async throws -> Data

    func handleCreateInstance(payload: Data) async throws -> Data
}

public extension PluginHandler {
    func handleCreateInstance(payload: Data) async throws -> Data {
        throw PluginRuntimeError.notStateful(plugin: String(describing: type(of: self)))
    }
}

public enum PluginRuntimeError: Error {
    case unknownMethod(plugin: String, method: String)
    case unknownInstance(UInt64)
    case notStateful(plugin: String)
}

public protocol HandleReleaser: AnyObject {
    func releaseNativeHandle(_ handleId: UInt64)
}

File: runtime/android/src/main/java/dev/istmo/runtime/PluginHandler.kt
package dev.istmo.runtime

import kotlinx.coroutines.flow.Flow

typealias NativeHandleId = Long

/**
 * Result of [PluginHandler.handleCall]. Unary methods produce
 * [Unary], stream methods (declared with `#[stream]` on the Rust
 * side) produce [Stream] — the runtime consumes each `ByteArray`
 * emitted by the flow as a `Frame::Event` back to the caller, and
 * closes the wire stream when the flow completes or errors.
 */
sealed class PluginResult {
    class Unary(val bytes: ByteArray) : PluginResult()
    class Stream(val flow: Flow<ByteArray>) : PluginResult()
}

interface PluginHandler {
    suspend fun handleCall(instanceId: Long, method: String, payload: ByteArray): PluginResult

    suspend fun handleCreateInstance(payload: ByteArray): ByteArray =
        throw PluginException(ByteArray(0))
}

class PluginException(val payload: ByteArray) : RuntimeException()

class BackendException(val error: Any) : RuntimeException()

interface HandleReleaser {
    fun releaseNativeHandle(handleId: Long)
}
